### PR TITLE
support reusing pods in pod evaluator

### PR DESCRIPTION
**Pull request by mengqiy**
Added a goroutine to manage the cache and handle the coordination.
Use to labels to retrieve old pods.
Set last use time in annotation.
Add GC to delete old pods.

I have manually tested it. Created the PR to get some early feedback.
I'm working on finding a way to test it without making the test setup too complex.

### CODE REVIEW

File: porch/func/internal/podevaluator.go
Comment (martinmaly) [inline porch/func/internal/podevaluator.go:85]: is the key for both maps the same?
If so, we could combine them into a single map, keyed by image and it contains the cached entrypoint info and the gRPC client to then call the function.

File: porch/func/internal/podevaluator.go
Comment (martinmaly) [inline porch/func/internal/podevaluator.go:87]: OK to add TODO for now and implement in a follow-up.
we should have a graceful shutdown for this goroutine.

File: porch/func/internal/podevaluator.go
Comment (martinmaly) [inline porch/func/internal/podevaluator.go:91]: indeed!

File: porch/func/internal/podevaluator.go
Comment (martinmaly) [inline porch/func/internal/podevaluator.go:144]: although we should probably check local cache first, and if not found, check for live pods (created by another runner), and if not found, create a new pod?

File: porch/func/internal/podevaluator.go
Comment (martinmaly) [inline porch/func/internal/podevaluator.go:162]: should we do this check before listing the pods above?

File: porch/func/internal/podevaluator.go
Comment (martinmaly) [inline porch/func/internal/podevaluator.go:153]: all of these operations will add latency to the function evaluation. We'll want a fast path to fetch gRPC client from a cache and make a call through it, and then secondary slow paths that handle cases when the client or pod don't exist.

As for last use update, we could also do it after the call (outside of the critical request/response path).

File: porch/func/internal/podevaluator.go
Comment (martinmaly) [inline porch/func/internal/podevaluator.go:126]: we'll need synchronization between the garbageCollector and the request handler goroutines

this will need to handle cases not only of concurrent access and updates, but also concurrent cache misses with subsequent pod spin-up.

Ideally we need logic such as:

```
gRPCClient, found := findCachedClientForFunction(image)
if !found {
  gRPCClient = coordinatePodSpinUpAndWakeMeWhenClientAvailable(image)
}
```

the coordinated spin-up should block all but one request as they wait for the pod to come up, while the running goroutine does the spin-up.

One way to do it is to have a goroutine dedicated to managing the image cache and goroutines handling the requests get the gRPC client(s) via channels. That way the synchronization may be easiest.

File: porch/func/internal/podevaluator.go
Comment (martinmaly) [inline porch/func/internal/podevaluator.go:298]: I'd recommend making this the 'slow path' of the request, should we get a cache miss on "which gRPC client is handling this function?" If we don't find the client we can then spin up the pod and get its client.

There is also potential for 'thundering herd' problem - if multiple clients call the same (uncached) function we'll start spinning up multiple pods in parallel.

Comment (martinmaly) [reply porch/func/internal/podevaluator.go:298]: We can do much of this as follow-ups

**Review by martinmaly — APPROVED** [review]
Some comments, but many of them can be addressed in follow-ups.

File: porch/func/internal/podevaluator.go
Comment (mengqiy) [reply porch/func/internal/podevaluator.go:126]: > we'll need synchronization between the garbageCollector and the request handler goroutines

Yes, you are right!

Re. coordinated spin-up, there are 2 levels of coordination we may need to spin up pods:
- Coordination between goroutines In one function runner
- Coordination between function runners

To achieve the later coordination, we may want to use k8s's coordination API: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#lease-v1-coordination-k8s-io

File: porch/func/internal/podevaluator.go
Comment (mengqiy) [reply porch/func/internal/podevaluator.go]: Done

File: porch/func/server/server.go
Comment (mengqiy) [reply porch/func/server/server.go]: Done

File: porch/func/internal/podevaluator.go
Comment (mengqiy) [reply porch/func/internal/podevaluator.go]: Made patching asynchronous

File: porch/func/internal/podevaluator.go
Comment (mengqiy) [reply porch/func/internal/podevaluator.go]: It should have be resolved now. PTAL

File: porch/func/internal/podevaluator.go
Comment (martinmaly) [inline porch/func/internal/podevaluator.go:345]: the language referring to ordering (here in few other places) seems out-of-date since (I assume) refactoring created independent functions.

Maybe `// Try to retrieve the pod.`

File: porch/func/internal/podevaluator.go
Comment (martinmaly) [inline porch/func/internal/podevaluator.go:461]: oh, I need to start using wait.PollImmediate. Nice!

**Review by martinmaly — APPROVED** [review]
Thank you! This is super cool

Feel free to handle the feedback in a follow-up PR(s); I'll leave it up to you which comments you address before merging and which after in follow-ups.

**Comment by mengqiy** [comment]
I will address the comments in a follow-up PR.

**Pull request merged**
